# HG changeset patch
# User David Spies <davidnspies@gmail.com>
# Date 1391401588 28800
#      Sun Feb 02 20:26:28 2014 -0800
# Node ID 7c0db9c4e4545aa188bcb1b3703286ccfc5b250c
# Parent  c2d1869a95ee3fd3a35e2a122175041bce3ef7b0
issymmetric.m: Fix OOM when run on large sparse matrices (bug #41426)

* issymmetric.m: Avoid using '(:)' which creates an index vector which may
exceed idx_vector_type when used on sparse matrices.

diff --git a/scripts/linear-algebra/issymmetric.m b/scripts/linear-algebra/issymmetric.m
--- a/scripts/linear-algebra/issymmetric.m
+++ b/scripts/linear-algebra/issymmetric.m
@@ -35,17 +35,18 @@ function retval = issymmetric (x, tol = 
 
   if (nargin < 1 || nargin > 2)
     print_usage ();
   endif
 
   retval = isnumeric (x) && issquare (x);
   if (retval)
     if (tol == 0)
-      retval = all ((x == x.')(:));
+      ## Handle large sparse matrices as well as full ones
+      retval = nnz (x != x.') == 0;
     else
       norm_x = norm (x, inf);
       retval = norm_x == 0 || norm (x - x.', inf) / norm_x <= tol;
     endif
   endif
 
 endfunction
 
@@ -54,16 +55,17 @@ endfunction
 %!assert (! issymmetric ([1, 2]))
 %!assert (issymmetric ([]))
 %!assert (issymmetric ([1, 2; 2, 1]))
 %!assert (! (issymmetric ("test")))
 %!assert (issymmetric ([1, 2.1; 2, 1.1], 0.2))
 %!assert (issymmetric ([1, 2i; 2i, 1]))
 %!assert (! (issymmetric ("t")))
 %!assert (! (issymmetric (["te"; "et"])))
+%!assert (issymmetric (speye (100000)))
 
 %!test
 %! s.a = 1;
 %! assert (! issymmetric (s));
 
 %!error issymmetric ([1, 2; 2, 1], 0, 0)
 %!error issymmetric ()
 
